### PR TITLE
Fix restore flow allowing a user to log in with an invalid mnemonic

### LINKED ISSUE #1416: The restore flow allows log in with an invalid mnemonic
**Describe the bug**
A mnemonic like "dad dad dad dad dad dad dad dad dad dad dad dad" is invalid (invalid checksum), but the word "dad" is in the dictionary.

The restore workflow doesn't validate the mnemonic, so you are logged in with a mnemonic you can't use again by trying to unlock the wallet.

### PR DESCRIPTION
**Pull request by deuch13**
Resolves #1416

### CODE REVIEW

File: web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js [hunk 1]
```diff
@@ -1,18 +1,295 @@
...
+	afterEach(async () => {
+		cleanup();
+		clearAndInitSpy.mockClear();
+		getWalletSpy.mockClear();
+		gotoSpy.mockClear();
+		settingsStore.reset();
+		settingsResetSpy.mockClear();
+		walletGetPsksSpy.mockClear();
+		walletStore.reset();
+		Object.defineProperty(window, "navigator", {
```
Comment (ascartabelli) [inline web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:57]: You don't need to do this after each test: you can set it up before with something alone the lines of:

```js
const readTextMock = vi.fn().mockResolvedValue(mnemonic);

Object.assign(window.navigator, { clipboard: { readText: readTextMock } } );
```

Then in the `afterEach` block you just clear the mock.

File: web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js [hunk 2]
```diff
@@ -1,18 +1,295 @@
...
+	const mnemonic = generateMnemonic();
+	const invalidMnemonic = "dad dad dad dad dad dad dad dad dad dad dad dad";
+	const pwd = "passwordpassword";
+	const loginInfo = await encryptMnemonic(mnemonic, pwd);
+	const seed = getSeedFromMnemonic(mnemonic);
+	const userId = (await new Wallet(seed).getPsks())[0];
+	const getWalletSpy = vi.spyOn(walletService, "getWallet");
+	const gotoSpy = vi.spyOn(appNavigation, "goto");
+	const settingsResetSpy = vi.spyOn(settingsStore, "reset");
+	const originalNavigator = { ...window.navigator };
```
Comment (ascartabelli) [inline web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:45]: No need to do this: see below

File: web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js [hunk 3]
```diff
@@ -1,18 +1,295 @@
...
 
 		expect(container.firstChild).toMatchSnapshot();
 	});
+
+	it("should render the Mnemonic Authenticate step after accepting the Existing Wallet Notice and the Terms of Service", async () => {
+		settingsStore.update(setKey("userId", userId));
+
+		const { container, getByRole } = render(Restore);
+
+		const noticeButton = getByRole("button", { name: "Proceed" });
```
Comment (ascartabelli) [inline web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:91]: In general, I don't think you need to store the button in a variable if you don't use it later on. You can just:

```js
await fireEvent.click(getByRole("button", { name: "Proceed" }));
```

File: web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js [hunk 4]
```diff
@@ -1,18 +1,295 @@
...
+			clipboard: {
+				read: vi.fn(),
+				readText: () => Promise.resolve(mnemonic),
+				write: vi.fn(),
+				writeText: vi.fn(),
+				addEventListener: vi.fn(),
+				dispatchEvent: vi.fn(),
+				removeEventListener: vi.fn()
+			}
+		});
```
Comment (ascartabelli) [inline web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:116]: You don't need all of this: see above.

File: web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js [hunk 5]
```diff
@@ -1,18 +1,295 @@
...
+			clipboard: {
+				read: vi.fn(),
+				readText: () => Promise.resolve(invalidMnemonic),
+				write: vi.fn(),
+				writeText: vi.fn(),
+				addEventListener: vi.fn(),
+				dispatchEvent: vi.fn(),
+				removeEventListener: vi.fn()
+			}
+		});
```
Comment (ascartabelli) [inline web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:148]: You don't need all of this: see above.

In this case you just need:

```js
readTextMock.mockResolvedValueOnce(invalidMnemonic);
```

File: web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js [hunk 6]
```diff
@@ -1,18 +1,295 @@
...
+			clipboard: {
+				read: vi.fn(),
+				readText: () => Promise.resolve(mnemonic),
+				write: vi.fn(),
+				writeText: vi.fn(),
+				addEventListener: vi.fn(),
+				dispatchEvent: vi.fn(),
+				removeEventListener: vi.fn()
+			}
+		});
```
Comment (ascartabelli) [inline web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:182]: You don't need all of this: see above.

File: web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js [hunk 7]
```diff
@@ -1,18 +1,295 @@
...
```
Comment (ascartabelli) [inline web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:239]: You don't need all of this: see above.

**Review by ascartabelli — CHANGES_REQUESTED** [review]
See my notes about the check in the code and the tests.

The thing that bothers me though is that the user can't proceed with an invalid mnemonic like `DAD DAD DAD DAD DAD DAD DAD DAD DAD DAD DAD DAD`, but without any explanation at all.

As the words came from the suggestions I think an error message is needed to explain the situation.

File: web-wallet/src/routes/(welcome)/setup/restore/MnemonicAuthenticate.svelte [hunk 8]
```diff
@@ -10,7 +11,8 @@
 	/** @type {string[]} */
 	export let enteredMnemonicPhrase = [];
 
-	$: isValid = enteredMnemonicPhrase.filter(word => word !== "").length === wordLimit;
+	$: isValid = enteredMnemonicPhrase.filter(word => word !== "").length === wordLimit
```
[the remark this answers is not in the document]
Comment (deuch13) [reply web-wallet/src/routes/(welcome)/setup/restore/MnemonicAuthenticate.svelte:14]: Indeed, that is not needed anymore if we have the `validateMnemonic()` function

File: web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js (hunk 1 above)
Comment (deuch13) [reply web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:57]: Have you tested this approach and got the same results?

Comment (ascartabelli) [reply web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:57]: even better results! (the paste button is not in the snapshot now, for example)

In this case you'll need to use `act` to have the mock promise settled though (or await a `tick`).

Comment (deuch13) [reply web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js:57]: > even better results! (the paste button is not in the snapshot now, for example)
> 
> In this case you'll need to use `act` to have the mock promise settled though (or await a `tick`).

Why should the paste button not be in the snapshot? It is part of the markup and should be shown.

Comment (ascartabelli) [reply web-wallet/src/routes/(welcome)/setup/restore/__tests__/page.spec.js]: > Why should the paste button not be in the snapshot? It is part of the markup and should be shown.

Yes, but it isn't there in your current snaphshot. With the change above it'll be there from the start.

**Review by ascartabelli — APPROVED** [review]
Not sure the "toast" is the best way to convey the message, to be honest, but will do for now.

**Pull request merged**